### ISSUE TITLE
Error: Unsupported argument

### ISSUE THREAD

**Issue by youngd24** [josenk/terraform-provider-esxi#46]
Using your configuration example:

provider "esxi" {
  esxi_hostname      = "esx01"
  esxi_hostport      = "22"
  esxi_username      = "root"
  esxi_password      = "xxx"
}

resource "esxi_guest" "vmtest" {
  guest_name         = "vmtest"
  disk_store         = "datastore1"
  ovf_source         = "/software/VM_Images/OVA/DEBIAN_9.7_X64_MINIMAL_LVM/DEBIAN_9.7_X64_MINIMAL_LVM.ova"

  network_interfaces = [
    {
      virtual_network = "VLAN47"
    },
  ]
}

ansible@util01:~/tf$ terraform version
Terraform v0.12.0-dev

terraform plan gets:

Error: Unsupported argument

  on main.tf line 13, in resource "esxi_guest" "vmtest":
  13:   network_interfaces = [

An argument named "network_interfaces" is not expected here. Did you mean to
define a block of type "network_interfaces"?

Any help would be appreciated, new(ish) to Terraform.

**Comment by josenk** [comment]
You're using a development version of terraform.   Terraform 0.12 has many changes that are not yet implemented here.  

You can get past that error message by using the upgrade tool, but this plugin is still not yet compatible with 0.12, so terraform plan/apply will still fail...
```
# terraform 0.12upgrade
```

If you want this plugin to work, you'll need to use the 0.11 version for now.

**Comment by youngd24** [comment]
Looks like the "go get -u -v github.com/hashicorp/terraform" in the instructions pulls the 0.12.0-dev version. Pulled the download from Hashicorp and will see if I can make that work.

**Comment by youngd24** [comment]
The downloaded version direct from Hashicorp worked and "go get -u -v github.com/hashicorp/terraform" always gets the 0.12.0-dev version right now.

**Comment by josenk** [comment]
Thanks for the update.   I'm working on the plugin to make it compatible with 0.12.  I'll update this issue when it's ready.

**Comment by josenk** [comment]
I did some initial work, getting this plugin working on Terraform 0.12...   

>https://github.com/josenk/terraform-provider-esxi/releases/tag/1.4.3

I have 1 issue I'm trying to figure out, but I believe it's workable right now...  There is a known issue that you must have virtual_disks block even if you don't have additional virtual disks to be added to your guest.  Just add an empty block.

```
virtual_disks {
}
```

   Any beta testers out there, please let me know how it's working out...

**Comment by XenonPK** [comment]
@josenk, the `userdata.encoding` configuration is causing issues:

```Error: Reference to undeclared resource

  on ../modules/vm/main.tf line 36, in resource "esxi_guest" "vm":
  36:     userdata.encoding = "gzip+base64"

A managed resource "userdata" "encoding" has not been declared in vm.```

**Comment by silasb** [comment]
@XenonPK Can you either post your terraform files or come up with a simplified version so we can reproduce?

**Comment by josenk** [comment]
I believe I saw this problem when using terraform 0.12 and trying to set userdata encoding.

**Comment by XenonPK** [comment]
@silasb This happens with the configuration provided in the examples:

```
#########################################
#  ESXI Guest resource
#########################################
resource "esxi_guest" "Default" {
  guest_name         = "${var.vm_hostname}"
  disk_store         = "${var.disk_store}"

  clone_from_vm      = "Templates/centos7"

  network_interfaces = [
    {
      virtual_network = "${var.virtual_network}"
    },
  ]

  guestinfo = {
    userdata.encoding = "gzip+base64"
    userdata = "${base64gzip(data.template_file.Default.rendered)}"
  }

}
```
The configuration provided in [this](https://github.com/josenk/terraform-provider-esxi/blob/master/examples/05%20CloudInit%20and%20Templates/main.tf) file triggers the error described.
With terraform 0.11.x it works as intended

**Comment by silasb** [comment]
Okay, I think I've figured it out.  I think this has to do with Go Modules, at least that's the only way I can get it working again.

Per https://www.terraform.io/docs/extend/terraform-0.12-compatibility.html

> Terraform Core is now using Go Modules for dependency management and vendoring, so we strongly recommend using Go Modules for dependency management in provider codebases too, which allows the go tool to automatically understand transitive dependencies and upgrade other required packages accordingly.

I think we should switch to using modules to better align with the Terraform SDK.

**Comment by josenk** [comment]
Agreed.

**Comment by josenk** [comment]
Should be fixed in v1.5.0
>https://github.com/josenk/terraform-provider-esxi/releases/tag/1.5.0

**Comment by DimDroll** [comment]
Can confirm.
Used this template: https://github.com/josenk/terraform-provider-esxi/blob/master/examples/01%20Simple%20Guest/main.tf
And terraform 0.12.5, works for me.
